### PR TITLE
Objective C: use C++ exceptions on MinGW+GNUstep

### PR DESCRIPTION
**Pull request by qmfrederik**
The GNUstep Objective C runtime (libobjc2) is adding support for the GNU ABI on Windows (more specifically, MinGW).  The libobjc2 runtime uses C++ exceptions in that configuration; this PR updates clang to act accordingly.

The corresponding change to libobjc2 is here: https://github.com/gnustep/libobjc2/pull/267

### CODE REVIEW

File: clang/lib/CodeGen/CGException.cpp [hunk 1]
```diff
@@ -156,7 +154,11 @@ static const EHPersonality &getObjCPersonality(const TargetInfo &Target,
   case ObjCRuntime::WatchOS:
     return EHPersonality::NeXT_ObjC;
   case ObjCRuntime::GNUstep:
-    if (L.ObjCRuntime.getVersion() >= VersionTuple(1, 7))
+    if (T.isOSCygMing())
```
Comment (qmfrederik) [inline clang/lib/CodeGen/CGException.cpp:157]: I moved the check for `isWindowsMSVCEnvironment` inside the `GNUstep` case, as on Windows we'll be using `MSVC_CxxFrameHandler3` when compiling with GNUstep + msvc; `GNU_CPlusPlus_SEH` when compiling with GNUstep + mingw and `GNU_ObjC_SEH` when compiling iwth GCC + mingw.

Comment (qmfrederik) [reply clang/lib/CodeGen/CGException.cpp:157]: OK - this broke some tests, so I'll revert that.

File: clang/lib/CodeGen/CGObjCGNU.cpp [hunk 2]
```diff
@@ -2210,7 +2219,12 @@ CGObjCGNU::CGObjCGNU(CodeGenModule &cgm, unsigned runtimeABIVersion,
 
   // void objc_exception_throw(id);
   ExceptionThrowFn.init(&CGM, "objc_exception_throw", VoidTy, IdTy);
-  ExceptionReThrowFn.init(&CGM, "objc_exception_throw", VoidTy, IdTy);
+  if ((CGM.getTarget().getTriple().isOSCygMing() &&
+       isRuntime(ObjCRuntime::GNUstep, 2))) {
+    ExceptionReThrowFn.init(&CGM, "objc_exception_rethrow", VoidTy, IdTy);
+  } else {
```
Comment (davidchisnall) [inline clang/lib/CodeGen/CGObjCGNU.cpp:2225]: It looks as if both branches of the if and else here do the same thing?

Comment (qmfrederik) [reply clang/lib/CodeGen/CGObjCGNU.cpp:2225]: On MinGW + objc2, `objc_exception_rethrow` is called; on other targets, `objc_exception_throw` is used (**re**throw vs throw).  I find it odd we were using `objc_exception_throw` to *re*throw an exception but decided to change it for MinGW+objc2 only.

Comment (davidchisnall) [reply clang/lib/CodeGen/CGObjCGNU.cpp:2225]: Ah, sorry, I didn't see the `re` bit.

File: clang/lib/CodeGen/CGException.cpp [hunk 3]
```diff
@@ -210,7 +212,10 @@ static const EHPersonality &getObjCXXPersonality(const TargetInfo &Target,
     return getObjCPersonality(Target, L);
 
   case ObjCRuntime::GNUstep:
-    return EHPersonality::GNU_ObjCXX;
+    if (Target.getTriple().isOSCygMing())
+      return EHPersonality::GNU_CPlusPlus_SEH;
+    else
+      return EHPersonality::GNU_ObjCXX;
```
Comment (compnerd) [inline clang/lib/CodeGen/CGException.cpp:218]: This might be nicer as a ternary return.

File: clang/lib/CodeGen/CGObjCGNU.cpp [hunk 4]
```diff
@@ -819,10 +819,19 @@ class CGObjCGNUstep : public CGObjCGNU {
...
+          isRuntime(ObjCRuntime::GNUstep, 2)) {
+        llvm::Type *VoidTy = llvm::Type::getVoidTy(VMContext);
+        // void *__cxa_begin_catch(void *e)
+        EnterCatchFn.init(&CGM, "__cxa_begin_catch", PtrTy, PtrTy);
+        // void __cxa_end_catch(void)
+        ExitCatchFn.init(&CGM, "__cxa_end_catch", VoidTy);
+        // void objc_exception_rethrow(void*)
+        ExceptionReThrowFn.init(&CGM, "__cxa_rethrow", PtrTy);
+      } else if (usesSEHExceptions) {
+        llvm::Type *VoidTy = llvm::Type::getVoidTy(VMContext);
```
Comment (compnerd) [inline clang/lib/CodeGen/CGObjCGNU.cpp:832]: Why not hoist this?  It is used across multiple branches.

File: clang/lib/CodeGen/CGObjCGNU.cpp [hunk 5]
```diff
@@ -3993,7 +4009,9 @@ void CGObjCGNU::EmitThrowStmt(CodeGenFunction &CGF,
     ExceptionAsObject = CGF.ObjCEHValueStack.back();
     isRethrow = true;
   }
-  if (isRethrow && usesSEHExceptions) {
+  if (isRethrow &&
+      (usesSEHExceptions || (CGM.getTarget().getTriple().isOSCygMing() &&
+                             isRuntime(ObjCRuntime::GNUstep, 2)))) {
```
Comment (compnerd) [inline clang/lib/CodeGen/CGObjCGNU.cpp:4014]: Does it make sense to introduce a `usesCxxExceptions` variable?

File: clang/lib/CodeGen/CGObjCGNU.cpp (hunk 4 above)
Comment (qmfrederik) [reply clang/lib/CodeGen/CGObjCGNU.cpp]: Good remark, thanks.  Fixed this.

File: clang/lib/CodeGen/CGObjCGNU.cpp (hunk 5 above)
Comment (qmfrederik) [reply clang/lib/CodeGen/CGObjCGNU.cpp]: Yes, that makes the intent clearer.  I changed this.

**Comment by qmfrederik** [comment]
@compnerd Thanks for you feedback, I addressed your comments.

**Comment by qmfrederik** [comment]
@compnerd @davidchisnall I've addressed the feedback and CI is (mostly) green.  The build failure on Windows seems unrelated:


```
# Removing c:/ws\src
# Creating "c:/ws\src"
> git clone -v -- https://github.com/llvm-premerge-tests/llvm-project.git .
Cloning into '.'...
remote: Internal Server Error
fatal: unable to access 'https://github.com/llvm-premerge-tests/llvm-project.git/': The requested URL returned error: 500
```

**Comment by davidchisnall** [comment]
I'd like CI to be green.  Hopefully the failure is transient, but I don't seem to be able to kick off a rerun.  Maybe once others start passing, you can squash the two commits and force push to restart it?

**Comment by qmfrederik** [comment]
Looks like there was an intermittent issue with GitHub, which should be resolved now.  I've squashed and pushed.  Fingers crossed!

**Comment by qmfrederik** [comment]
CI looks green, let me know if there's anything else you need from me on this PR!

File: clang/lib/CodeGen/CGObjCGNU.cpp [hunk 6]
```diff
@@ -2210,7 +2219,11 @@ CGObjCGNU::CGObjCGNU(CodeGenModule &cgm, unsigned runtimeABIVersion,
 
   // void objc_exception_throw(id);
   ExceptionThrowFn.init(&CGM, "objc_exception_throw", VoidTy, IdTy);
-  ExceptionReThrowFn.init(&CGM, "objc_exception_throw", VoidTy, IdTy);
+  if (usesCxxExceptions) {
+    ExceptionReThrowFn.init(&CGM, "objc_exception_rethrow", VoidTy, IdTy);
```
Comment (compnerd) [inline clang/lib/CodeGen/CGObjCGNU.cpp:2223]: I would say that a ternary to select between the exception model and identify the name would be better as you are just selecting the label.

**Review by compnerd — APPROVED** [review]

**Comment by davidchisnall** [comment]
It looks as if three tests are failing in CI:

  Clang :: CodeGenObjC/2007-04-03-ObjcEH.m
  Clang :: CodeGenObjC/exceptions-personality.m
  Clang :: Coverage/codegen-gnu.m

**Pull request merged**